Proxy TAG requests through Auth

jakule opened the pull request:
Related `e` PR https://github.com/gravitational/teleport.e/pull/2663

jakule committed:
Proxy TAG requests through Auth

jakule committed:
Integrate Access Graph feature and optimize services

This commit entails the integration of the Access Graph feature into the existing service configuration. This feature is manifested through an 'IsEnabled' method that checks the state of the access graph. Additionally, the auth service logic now includes a flag for enabling access graph. The services have been further optimized with adjustments geared towards improving performance and readability of the code.

jakule committed:
Move TAG enabled flag from proxy to user setting/auth context

jakule committed:
Swap a pointer to AccessGraph config with a value.

jakule committed:
Add 'TestNewAccessGraph' function to 'useracl_test.go' file.

The function 'TestNewAccessGraph' has been added to run tests checking the configuration of Access Graph. Furthermore, a small change is made for the 'KindAccessGraph' comment in 'constants.go' to provide more accurate information.

zmb3 commented on api/proto/teleport/legacy/client/proto/authservice.proto:
```diff
@@ -524,6 +524,8 @@ message Features {
   string CustomTheme = 22 [(gogoproto.jsontag) = "custom_theme,omitempty"];
   // IdentityGovernance holds the Identity Governance feature settings (Access Monitoring)
   bool IdentityGovernance = 23 [(gogoproto.jsontag) = "identity_governance,omitempty"];
+  // AccessGraph holds the Access Graph feature settings.
```
nit: it doesn't hold "settings" - it holds a single boolean value

```suggestion
  // AccessGraph indicates whether Teleport Access Graph is enabled.
```

justinas commented on lib/auth/auth.go:
```diff
@@ -5261,6 +5265,8 @@ func (a *Server) Ping(ctx context.Context) (proto.PingResponse, error) {
 		features.IdentityGovernance = a.accessMonitoringEnabled
 	}
 
+	features.AccessGraph = a.accessGraphEnabled
```
I would vote for reusing the pattern for `Plugins`:

Can't say I particularly like this, as this will result in confusingly desynced `Features()` in Auth server itself vs what is returned to the proxy.

I would suggest reusing this pattern:

https://github.com/gravitational/teleport.e/blob/e5403beda326e90e8562c8a73831087ffb673573/tool/modules/modules.go#L111-L139

and then call `EnableAccessGraph()` from `auth/plugin.go` in `e` (this will mesh well with the "is team plan" check I'm adding).

Adding a new method `EnableAccessGraph()` makes the merge a bit complicated, since we need to add it in `e` first, and only then in OSS (so that the interface is satisfied). We could do a type-cast i.e. `modules.GetModules().(*enterpriseModules)` for now in `e` and add a `TODO`, I will fix it my PR.

jakule replied:
Changed

jakule committed:
Implement Access Graph feature across various files

The Access Graph feature has been incorporated in 'module.go', 'authservice.proto', 'auth.go', and 'plugin.go' files. A method 'EnableAccessGraph' is also introduced in 'module.go' to allow the usage of this feature. Descriptions of Access Graph have been adjusted across files to explicitly state the feature's enablement.

jakule committed:
Fix tests

justinas commented on lib/auth/auth.go:
```diff
@@ -803,6 +804,9 @@ type Server struct {
 	// accessMonitoringEnabled is a flag that indicates whether access monitoring is enabled.
 	accessMonitoringEnabled bool
 
+	// accessGraphEnabled is a flag that indicates whether access graph is enabled.
```
This is now unused I think

jakule replied:
Removed.

justinas commented on lib/auth/init.go:
```diff
@@ -270,6 +270,9 @@ type InitConfig struct {
 
 	// AccessMonitoringEnabled is true if access monitoring is enabled.
 	AccessMonitoringEnabled bool
+
```
This is also unused now

jakule replied:
Removed

jakule committed:
Remove unused access graph feature code

This commit removes all references and uses of the access graph feature from the codebase. It also updates corresponding tests to reflect this removal. It also adds EnableAccessGraph() method to testing module to fix test compilation.
